api_app/analyzers_manager/file_analyzers/cape_sandbox.py

```diff
@@ -0,0 +1,232 @@
+        ):
+
+            """
+                The above response is only returned when a sample that has been already
+                uploaded once is uploaded again.
+
+                If it has been then we can just check it's
+                report by querying the CAPESandbox API with the md5 hash of
+                the file.
+
+                If it exists in their database and is readable by us,
+                the following code further fetches it's information.
+                response_json in this case should look somewhat like this:
+
+            {
+                'error': True,
+                'error_value': 'Error adding task to database',
+                'errors': [{
+                    'filename.exe':
+                        'Not unique, as unique option set on submit or in conf/web.conf'
+                }]
+            }
+            """
+
+            logger.info(
+                f"[INFO] (Job: {self.job_id} -> "
+                "File uploaded is already present in the database. "
+                "Querying its information through it's md5 hash.."
+            )
+
+            status_id = self.__search_by_md5()
+            gui_report_url = self.__base_url + "/submit/status/" + status_id
+            report_url = (
+                self.__base_url + "/apiv2/tasks/get/report/" + status_id + "/json"
+            )
+            to_respond["result_url"] = gui_report_url
+
+            try:
+                final_request = self.__session.get(
+                    report_url,
+                )
+            except requests.RequestException as e:
+                raise AnalyzerRunException(e)
+
+            final_json = final_request.json()
+            to_respond["response"] = final_json
+
+            response_json = to_respond
+
+        return response_json
+
+    def __search_by_md5(self) -> str:
+        db_search_url = self.__base_url + "/apiv2/tasks/search/md5/" + self.md5
+
+        try:
+            q = self.__session.get(db_search_url)
+            q.raise_for_status()
+
+        except requests.RequestException as e:
+            raise AnalyzerRunException(e)
+
+        data_list = q.json().get("data")
+        if data_list:
+            status_id_int = data_list[0].get("id")
+            status_id = str(status_id_int)
+            return status_id
+
+        raise AnalyzerRunException(
+            "'data' key in response isn't populated as expected."
+        )
+
+    def __poll_for_result(
+        self,
+        task_id,
+    ) -> dict:
+        status_api = self.__base_url + "/apiv2/tasks/status/" + str(task_id)
+        for i in range(self.max_tries):
+            logger.info(
+                f"[POLLING] (Job: {self.job_id} -> "
+                f"CAPEsandbox __poll_for_result #{i + 1}/{self.max_tries}"
+            )
+            try:
+                r = self.__session.get(status_api)
+                if r.status_code == 429:
```

Comment:
a question: so either you sleep or you raise an exception? So where is the "normal and right" case with the 200 status code?

Reply:
plus, why there is a sleep here and then you do not continue the loop? I mean, the logic is really difficult to follow in this for clause. You also do a sleep later. Please manage those sleeps with a single custom exception that, when is triggered, it performs the sleep and continues the loop

Reply:
> Please manage those sleeps with a single custom exception that, when is triggered, it performs the sleep and continues the loop

Good suggestion. Will do.

Reply:
Now is better than before but there are still some problems. My point was: when something happens and you need to continue the loop, you should handle it appropriately at that specific time.

For instance, here (with the 429 status code) you want to sleep and continue the loop but...it does not happen exactly here, while it should. On the contrary you wait for the end of the code to do that. That is not how it should work. You should trigger a Custom Exception here that would bring you directly at the end of the code of the loop where there is a log + sleep.

Same thing for what happens after the `RequestException`. Also there you don't manage the problem directly. In both cases, you should raise the same Custom Exception that would bring you here:
```
            logger.info(
                f"Job: {self.job_id} -> "
                f"Analysis status: {data}. "
                f"Sleeping for {self.poll_distance} seconds before next attempt."
            )
            time.sleep(self.poll_distance)
```

I mean, you need to separate the part where you handle the loop for the first request and the part where you handle the results and do not mix them